cpp/src/reductions/scan.cu

```diff
@@ -47,29 +48,30 @@ struct Scan {
     }
 };
 
-#define SCAN_IMPL(F, T)                                                       \
-gdf_error gdf_prefixsum_##F(gdf_column *inp, gdf_column *out, int inclusive) {\
-    GDF_REQUIRE( inp->size == out->size, GDF_COLUMN_SIZE_MISMATCH );          \
-    GDF_REQUIRE( inp->dtype == out->dtype, GDF_UNSUPPORTED_DTYPE );           \
-    GDF_REQUIRE( !inp->valid || !inp->null_count, GDF_VALIDITY_UNSUPPORTED ); \
-    GDF_REQUIRE( !out->valid || !out->null_count, GDF_VALIDITY_UNSUPPORTED ); \
-    return Scan<T>::call((const T*)inp->data, (T*)out->data, inp->size,       \
-                         inclusive);                                          \
-}
-
-
-SCAN_IMPL(i8,  int8_t)
-SCAN_IMPL(i32, int32_t)
-SCAN_IMPL(i64, int64_t)
+struct PrefixSumDispatcher {
+    template <typename T,
+        typename std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
+        gdf_error operator()(gdf_column *inp, gdf_column *out,
+            int inclusive) {
+        GDF_REQUIRE(inp->size == out->size, GDF_COLUMN_SIZE_MISMATCH);
+        GDF_REQUIRE(inp->dtype == out->dtype, GDF_UNSUPPORTED_DTYPE);
+        GDF_REQUIRE(!inp->valid || !inp->null_count, GDF_VALIDITY_UNSUPPORTED);
+        GDF_REQUIRE(!out->valid || !out->null_count, GDF_VALIDITY_UNSUPPORTED);
+        return Scan<T>::call((const T*)inp->data, (T*)out->data, inp->size,
```

Comment:
Prefer `static_cast` vs. a C-style cast.

Reply:
I prefer `reinterpret_cast`. Is it good?

Reply:
The problem with using a C-style cast is that it's doing a `reinterpret_cast`, which is unsafe if you don't really need it. It's a much bigger hammer than `static_cast`, which is safer because it prevents you from doing some unsafe things. That's why it's always best to prefer `static_cast` and only use `reinterpret_cast` when you absolutely need it.

Reply:
Here's a really good SO post on the issue https://stackoverflow.com/a/332086

Reply:
@kovaltan : Using reinterpret_cast's often makes compilers give up on trying various optimizations.

Reply:
Undserstand. Will use static_cast at next patch in next week.